src/StoreApi/Routes/Checkout.php

```diff
@@ -139,71 +139,86 @@ protected function get_route_update_response( \WP_REST_Request $request ) {
-			// for setting initial password.
-			try {
-				$create_account = Package::container()->get( CreateAccount::class );
-				$create_account->from_order_request( $request );
-				$order_object->set_customer_id( get_current_user_id() );
-			} catch ( Exception $error ) {
-				$this->handle_error( $error );
-			}
+		// Validate payment method before proceeding.
+		if ( $order->needs_payment() ) {
+			// Will throw and return 400 if payment method missing or invalid.
+			$this->get_request_payment_method_id( $request );
 		}
-		// If any form fields were posted, update the order.
-		$this->update_order_from_request( $order_object, $request );
 
-		// Check order is still valid.
-		$order_controller->validate_order_before_payment( $order_object );
+		// --
+		// 3. Process Customer
 
-		// Persist customer address data to account.
-		$order_controller->sync_customer_data_with_order( $order_object );
+		// Update order with customer details, and sign up a user account as necessary.
+		$this->process_customer( $order, $request );
+
+		// --
+		// 4. Validate Order
+
+		// Check order is still valid.
+		$order_controller = new OrderController();
+		$order_controller->validate_order_before_payment( $order );
 
 		/*
-		* Fire woocommerce_blocks_checkout_order_processed, should work the same way as woocommerce_checkout_order_processed
-		* But we're opting for a new action because the original ones attaches POST data.
-		* NOTE: this hook is still experimental, and might change or get removed.
-		* @todo: Document and stabilize __experimental_woocommerce_blocks_checkout_order_processed
+		* Action: woocommerce_blocks_checkout_order_processed (experimental).
+		* This hook informs extensions that $order has completed processing and is ready
+		* for payment.
+		*
+		* This is similar to existing core hook woocommerce_checkout_order_processed.
+		* We're using a new action:
+		* - To keep the interface focused (only pass $order, not passing request data).
+		* - This also explicitly indicates these orders are from checkout block/StoreAPI.
+		*
+		* Status: experimental; may change or be removed.
+		* Introduced: WooCommerce Blocks 3.8.0.
+		* https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3238
 		*/
-		do_action( '__experimental_woocommerce_blocks_checkout_order_processed', $order_object );
+		// Save the order so the param and database match (extensions may access DB).
+		$order->save();
+		do_action( '__experimental_woocommerce_blocks_checkout_order_processed', $order );
+
+		// --
+		// 5. Process Payment
 
-		if ( ! $order_object->needs_payment() ) {
-			$payment_result = $this->process_without_payment( $order_object, $request );
+		if ( ! $order->needs_payment() ) {
+			$payment_result = $this->process_without_payment( $order, $request );
 		} else {
-			$payment_result = $this->process_payment( $order_object, $request );
+			$payment_result = $this->process_payment( $order, $request );
```

Comment:
This is where you'd feed in the `$payment_method_id` if you implement my suggested changes above.